sdk/core/dma/dma_compartment.cc

```diff
@@ -0,0 +1,75 @@
+#pragma once
+
+#define MALLOC_QUOTA 0x100000
+
+#include <cheri.hh>
+#include <compartment-macros.h>
+#include <utils.hh>
+
+#include <thread_pool.h>
+
+using thread_pool::async;
+/**
+ *  creating a global heap below.
+ *  todo: we assume that this heap capability can serve 
+ *  for different compartments, irrespective of 
+ *  the origin because this is a driver code?
+*/
+DECLARE_AND_DEFINE_ALLOCATOR_CAPABILITY(dmaDriverHeap, 8144);
+
+SObjStruct *dmaHeap;
+
+int claim_dma(uint32_t *sourceAddress, uint32_t *targetAddress)
+{    
+    /**
+     *  create a heap capability for dma below. 
+     */
+    dmaHeap = STATIC_SEALED_VALUE(dmaDriverHeap);
+
+    /**
+     *  claim the memory 
+     */
+    
+    size_t sourceClaimSize = heap_claim(dmaHeap, sourceAddress);
```

Comment:
I wonder if this is sufficient grounds to want a `heap_claim(SObj heapCapability, void*... objects)` that amortizes the compartment call overhead.  Similarly for `heap_free()`.  @DavidChisnall?

Reply:
That’s a good idea. In particular, the call could atomically succeed or fail, which simplifies error handling for the caller. I think we’d want to bound the maximum number of objects that can be claimed in a single operation and we can’t used variadic an across compartment calls so it would have to pass an array and a length.